### PR TITLE
Add Notify Room Typing Event to Apps

### PR DESCRIPTION
**Pull request by Shailesh351**
## Usecase : 
To support Visitor Typing Event in Apps

## Related PRs :

Apps Engine: https://github.com/WideChat/Rocket.Chat.Apps-engine/pull/3
Salesforce App: https://github.com/WideChat/Apps.Salesforce.LiveAgent/pull/9

## Demo : https://youtu.be/DIckTccB5YU

### CODE REVIEW

File: app/livechat/server/lib/Livechat.js [hunk 1]
```diff
@@ -49,6 +51,21 @@ export const Livechat = {
 		},
 	}),
 
+	addTyingListener(rid, callback) {
```
Comment (dlassalle) [inline app/livechat/server/lib/Livechat.js]: You should do a search for 'Tying' --> Typing, looks like there are a couple spots where that shows up

**Comment by dlassalle** [comment]
I probably wont have time to test these until tomorrow morning, but in the meantime do you know if we are able to add our fork to the package.json like we do with the livechat widget?
https://github.com/WideChat/Rocket.Chat/blob/develop_pwa/package.json#L58
https://github.com/WideChat/Rocket.Chat/blob/develop_pwa/package.json#L134

For livechat we had to remove 'build' from the repo's gitignore

File: app/livechat/server/lib/Livechat.js (hunk 1 above)
Comment (Shailesh351) [reply app/livechat/server/lib/Livechat.js]: Thanks @dlassalle Fixed it ✅

**Comment by Shailesh351** [comment]
> I probably wont have time to test these until tomorrow morning, but in the meantime do you know if we are able to add our fork to the package.json like we do with the livechat widget?
> https://github.com/WideChat/Rocket.Chat/blob/develop_pwa/package.json#L58
> https://github.com/WideChat/Rocket.Chat/blob/develop_pwa/package.json#L134
> 
> For livechat we had to remove 'build' from the repo's gitignore

I will try at my side 👍

**Comment by Shailesh351** [comment]
@dlassalle It's possible with https://github.com/WideChat/Rocket.Chat.Apps-engine/pull/4/commits/ed5ee28262565588340e1187f251782b18fdad57
and https://github.com/WideChat/Rocket.Chat/commit/b830a385bb0c3f7f7096d4d722b3792bf2901449

**Comment by dlassalle** [comment]
Lets update the package.json here to use our apps-engine fork as well, assuming we'll merge the apps-engine PRs first

**Comment by Shailesh351** [comment]
@dlassalle Updated package.json dependency. We can merge after build succeed

**Review by dlassalle — APPROVED** [review]

**Comment by dlassalle** [comment]
@ear-dev this should be mergable

**Comment by ear-dev** [comment]
@dlassalle @Shailesh351 Oddly, I actually heard from Sing about this PR.......  it seems that somebody on the core team took a look at it and feels that it is the wrong approach.  Honestly, I had no idea that they were monitoring this fork for PRs, and I wish they had commented on the PR if so.....

image

image

Perhaps we should follow up with shiqui.mei  as Sing suggested?  It might also be good to know when they will get their own version in, so that maybe we will not have to do it ourselves?

**Review by ear-dev — CHANGES_REQUESTED** [review]
Lets figure out when the core team will have this in so that maybe we will not have to implement.  At least, lets figure out the recommended way of doing it.

**Comment by Shailesh351** [comment]
@ear-dev @dlassalle I will check from my side for any improvement and I'm also trying to connect with shique.mei. Will revert back as soon as I get any update from him.

Regarding `the bot is typing ...`, The feature of this PR is kind of reverse event. We want typing event in salesforce console, when **visitor is typing**. So when visitor is typing we send event to Salesforce App and in the App we call salesforce REST API for typing which will be reflected in salesforce console.

**Comment by d-gubert** [comment]
Hello folks!

After some more reading on my side, I've realized that this PR introduces the listening for user-typing event ONLY IN LIVECHAT ROOMS. This is not as problematic as having it listen for ALL user-typing events, but will be as resource intensive as the number of active concurrent livechat session you have. This is something inherent to the nature of the event (it does naturally get triggered a lot), so doing it efficiently is not trivial I think - one idea that comes to mind would be "batching" the triggers, with strategies like [debouncing/throttling](https://lodash.com/docs/4.17.15#debounce).

As to the feature we're working on right now, as @Shailesh351 pointed out, it is the other way around - we'll enable Rocket.Chat Apps to send custom user-typing notifications to clients, allowing automated bots to mimic their human counter-parts 🙂 

@ear-dev we're not monitoring this repo, @Shailesh351 shared this PR with us in our public #rocketchat-apps channel on the Open server, so we came here to take a look :)

**Comment by ear-dev** [comment]
Thanks for all those clarifications @d-gubert 

I'm happy that you took a look at the PR, it was just surprising to me because I was not aware of the context.  Please do feel free to comment on our PRs anytime you might stumble into them.....  we will always appreciate it. 

@Shailesh351 let me know if you think there is a more efficient implementation for this that you will try to work through.... we can discuss at our meeting tomorrow if you want.

File: app/livechat/server/api/lib/livechat.js [hunk 2]
```diff
@@ -71,9 +71,9 @@ export function findOpenRoom(token, departmentId) {
 	}
 
 	if (room) {
-		Livechat.addTypingListener(room._id, (username, typing) => {
+		Livechat.addTypingListener(room._id, _.debounce((username, typing) => {
 			Apps.triggerEvent(AppEvents.IRoomUserTyping, { roomId: room._id, username, typing });
-		});
+		}, 4000, true));
```
Comment (PrajvalRaval) [inline app/livechat/server/api/lib/livechat.js]: Hi @Shailesh351, I reviewed the PR, but the delay between the visitor starting to type and the indicator actually been shown to the agent is a little too much. Can we decrease that delay by a couple of seconds?

**Review by PrajvalRaval — CHANGES_REQUESTED** [review]

File: app/livechat/server/api/lib/livechat.js (hunk 2 above)
Comment (Shailesh351) [reply app/livechat/server/api/lib/livechat.js]: Hey @PrajvalRaval I tried some values but the number of triggers was high even if user is typing continuously with normal speed. 
In the LiveChat widget https://github.com/RocketChat/Rocket.Chat.Livechat/blob/48ae3dc366056c9985849719beed90a3fb87d50f/src/routes/Chat/container.js#L99 we have **4500 ms**.
May be we can go with 3000 or 3500. What do you think?

Comment (PrajvalRaval) [reply app/livechat/server/api/lib/livechat.js]: Can we do 2500? I think that will be a nice time, considering the average typing speed of a person.

Comment (Shailesh351) [reply app/livechat/server/api/lib/livechat.js]: Done @PrajvalRaval ✅

**Review by PrajvalRaval — APPROVED** [review]

**Comment by ear-dev** [comment]
Prajval approved.

**Pull request merged**